services/src/main/java/org/keycloak/authentication/authenticators/browser/ConditionalSpnegoAuthenticator.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright 2023 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+package org.keycloak.authentication.authenticators.browser;
+
+import org.jboss.logging.Logger;
+import org.keycloak.authentication.AuthenticationFlowContext;
+import org.keycloak.sessions.AuthenticationSessionModel;
+
+import java.util.Map;
+import java.util.regex.Pattern;
+
+/**
+ * A {@link SpnegoAuthenticator} that can conditionally skip SPNEGO authentication.
+ * <p>
+ * This Keycloak authenticator allows SPNEGO authentication to be skipped by explicit client request and optionally by
+ * request IP. Clients can request SPNEGO authentication be skipped via the URL prompt=login present in the client
+ * request. A Whitelist Regex Pattern can optionally be configured, and if configured, it is compared against the
+ * <i>X-Forwarded-For</i> header on auth requests to check for a match. If no match then no SPNEGO is skipped.
+ * </p>
+ * <p>
+ * It is often desirable to treat Intranet (internal) users with the convenience of Kerberos Single Sign On, while
+ * preventing Internet (external) users from possibly seeing a confusing NTLM prompt dialog.  It can also be useful to
+ * allow users to explicitly opt out of SPNEGO in order to provide a switch user functionality.
+ * </p>
+ * @author <a href="mailto:Ryan.Slominski@gmail.com">Ryan Slominski</a>
+ */
+public class ConditionalSpnegoAuthenticator extends SpnegoAuthenticator {
```

Comment:
It will be better to implement this as separate condition instead of extension of `SpnegoAuthenticator`.

I know we have already `ConditionalFormAuthenticatorFactory`, but that was introduced before we have conditions support for authentication flows. For the sake of better composing and ability to re-use the similar condition in other places too, the dedicated implementation of `ConditionalAuthenticator` is much better option.

Reply:
Ok, I'll look into `ConditionalAuthenticator`.   Just looking at the [docs](https://www.keycloak.org/docs/20.0.3/server_admin/index.html#conditional) it's a little confusing as it says:

> - If all executions evaluate as true, the Conditional sub-flow acts as Required.
> - If all executions evaluate as false, the Conditional sub-flow acts as Disabled.

I think this might be trying to say:

- If all executions evaluate as true, the Conditional sub-flow acts as Required.
- If __any__ execution evaluates as false, the Conditional sub-flow acts as Disabled.

Otherwise the case of some executions evaluate to true while others evaluate to false is left unexplained.

Also, a SPNEGO execution is usually configured as an `ALTERNATIVE`, meaning if it fails no problem just continue on to the next authenticator.

REQUIRED is defined in the docs as: 

> All Required elements in the flow must be successfully sequentially executed. The flow terminates if a required element fails.

Does that mean that if we use a `ConditionalAuthenticator` it is not possible to conditonally attempt SPNEGO, and if attempted, but fails, continue on to the next authenticator?  If possible, then maybe I'm confusing the definition of a REQUIRED subflow vs a REQUIRED execution?

Reply:
Yes, you're right. I think the docs should be updated as you suggested (regarding if *any* ...)

Regarding conditional subflow: I think that what you want to achieve probably can be achieved with something like replacing default Kerberos authenticator with something like:
```
- kerberos-subflow1 ALTERNATIVE
-- nested-conditional-subflow CONDITIONAL
--- condition1-reauthentication REQUIRED
--- condition2-xforwarded-headers REQUIRED
--- Kerberos ALTERNATIVE
```
Note that I can be wrong as I did not test it by myself. This configuration is a bit cumbersome and we plan to improve authentication flows by adding "Authentication policies" in the future, which will allow to configure them in the more intuitive way.

Reply:
Wow, that's a mess.  It looks like a the requirement type of `Conditional` unnecessarily complicates and de-normalizes.  The `Conditional` requirement should probably be deprecated and instead simply adding an execution type of Conditional subflow is enough to indicate you mean conditional subflow.  Having a separate requirement type is redundant and actually prevents you from indicating that the subflow is ALTERNATIVE.  Intuitively the only difference between a regular subflow and a conditional subflow would be that the regular one has ordered executions and a conditional subflow has ordered conditions and then one execution at the end that is only executed if the gauntlet of conditions are all passed.  It's actually not immediately clear how this is supposed to work.  It looks like this was defined in https://github.com/keycloak/keycloak/pull/7897

Reply:
Actually, it looks like the way conditional subflows work is if their condition fails then they can short-circuit the conditional subflow as if the entire subflow execution is DISABLED.  If their condition passes then execution is allowed to continue and any executions in the conditional flow behave as if they're in a regular flow and not in a subflow at all.  If that's true then the setup would look like:

```
- conditional-subflow CONDITIONAL
-- condition1-reauthentication
-- condition2-xforwarded-headers
-- Kerberos ALTERNATIVE
```

I don't think conditions have "requirements" - they just simply resolve to pass or fail.   I haven't tested any of this either and it's worrying that it's not immediately clear how this is supposed to work.

Reply:
The idea of a set of generic reusable conditions is good, but an ordered list of conditions should probably be an attribute of an execution, not a special subflow.

Reply:
Sorry about the confusion with the requirement on the conditions...

I think that for your configuration, the `conditional-subflow` will become `REQUIRED` in case that conditions evaluate to true. Hence it likely won't work correctly if you want it to use at same level with other things and achieve something like: cookie OR conditional-kerberos OR identity-provider-redirector OR forms-subflow

Hence why I think another "parent" alternative subflow might be needed to wrap the conditional-subflow... You can try and see.

I definitely agree that this stuff may require some more love to become more usable and intuitive to configure.

Reply:
I created a discussion about this [here](https://github.com/keycloak/keycloak/discussions/17152).   I'm not really sure how to proceed.    I'm leaning towards sticking with the current approach of extending SpnegoAuthenticator for now until the new way is improved.  It seems like the IP based Condition someone else is already working on may solve this partially.